pkgs/tools/misc/hdf4/default.nix

```diff
@@ -31,11 +31,12 @@ stdenv.mkDerivation rec {
     "-DHDF4_BUILD_WITH_INSTALL_NAME=OFF"
     "-DHDF4_ENABLE_JPEG_LIB_SUPPORT=ON"
     "-DHDF4_ENABLE_NETCDF=OFF"
-    "-DHDF4_ENABLE_SZIP_ENCODING=ON"
-    "-DHDF4_ENABLE_SZIP_SUPPORT=ON"
     "-DHDF4_ENABLE_Z_LIB_SUPPORT=ON"
     "-DHDF4_BUILD_FORTRAN=OFF"
     "-DJPEG_DIR=${libjpeg}"
+  ] ++ stdenv.lib.optionals (szip != null) [
+    "-DHDF4_ENABLE_SZIP_ENCODING=ON"
```

Comment:
When Hydra builds, it doesn't build unfree products. So, indeed what you need is for PostGIS to be free. 

You identified a dependency chain involving hdf4 and made the modification to *not* build it under some conditions. This condition is too strict, as I said before. 

You need to compute the license based on whether or not someone is building with szip encoding enabled (for which you need to introduce a parameter in the build script). The default for the variable needs to be set such that szip encoding is disabled. Additionally, this variable needs to be depend on the boolean variable NixOS uses allow/disallow the building of non-free code.

These same boolean variables need to be used to control the building of the software itself (i.e. the CMake variables). In *all* cases szip *decoding* should be compiled in.

If you do it like I described, Hydra will not set the nonfree flag and won't build the patented code. Although, it would be allowed to build the code, it just wouldn't be allowed to distribute the code (so, building it is useless). 

A user probably wouldn't want to infringe by accident, so they would get the szip compression free version by default from Hydra. So, PostGIS users not using szip won't notice anything. 

Now, there is another class of users, the commercial user with a license to use szip encoding (or someone who believes the patent is not valid (which, incidentally is *everyone* in Europe(!)) who could still build the software *including* all the features, in particular the szip encoding (which is `"-DHDF4_ENABLE_SZIP_ENCODING=ON"`). 

I hope this is now more clear to you.